### PR TITLE
Switch Java discovery samplegen to MVVM

### PR DESCRIPTION
**Pull request by saicheems**

### CODE REVIEW

**Comment by saicheems** [comment]
I've included a few baselines diffed against master, PTAL.

**Comment by vchudnov-g** [comment]
* @saicheems: Before we merge, please implement in a separate PR the equivalent of #542, but for MVVM
* @tcoffee-google tcoffee: Your close scrutiny of the baseline changes here will be extremely valuable

**Comment by saicheems** [comment]
I'd prefer to do so immediately after, as the media stuff depends on changes to apiary config

Otherwise I'll wait for him to merge that PR and make another followup

**Comment by tcoffee-google** [comment]
#542 has been merged.

**Comment by vchudnov-g** [comment]
@saicheems Let's do the media stuff first, so we have no expected regressions when we switch over.

**Comment by tcoffee-google** [comment]
Well, I can't view adexchangebuyer and compute on Github, so I'll have to format my comments separately. Stay tuned :-).

File: src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline [hunk 1]
```diff
@@ -1,871 +1,966 @@
...
+      if (response.getEntries() == null) {
         continue;
-
+      }
       for (LogEntry logEntry : response.getEntries()) {
         // TODO: Change code below to process each `logEntry` resource:
         System.out.println(logEntry);
       }
-
-      requestBody.setPageToken(response.getNextPageToken());
```
Comment (saicheems) [inline src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline]: Yeah, just noticed this, I have a pull ready for it.

File: src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline [hunk 2]
```diff
@@ -1,871 +1,966 @@
...
-      .build();
-
-
+  public void run() throws IOException, GeneralSecurityException {
+    // TODO: Assign values to desired fields of `requestBody`.
     ListLogEntriesRequest requestBody = new ListLogEntriesRequest();
-    // TODO: Assign values to desired fields of the `requestBody` object.
 
+    Logging loggingService = createService();
+    List request = loggingService.entries().list(requestBody);
```
Comment (saicheems) [inline src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline]: I still prefer it -
In some cases using the nearly fully qualified type can be very long, and that looks kinda odd and can cause line length issues

File: src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline [hunk 3]
```diff
@@ -1,871 +1,966 @@
 
-/**
+/*
```
Comment (saicheems) [inline src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline]: #556

File: src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline [hunk 4]
```diff
@@ -1,871 +1,966 @@
...
-
-    Logging loggingService = new Logging.Builder(httpTransport, jsonFactory, credential)
-      .setApplicationName("Google Cloud Platform Sample")
-      .build();
-
-
+  public void run() throws IOException, GeneralSecurityException {
+    // TODO: Assign values to desired fields of `requestBody`.
     ListLogEntriesRequest requestBody = new ListLogEntriesRequest();
-    // TODO: Assign values to desired fields of the `requestBody` object.
```
Comment (saicheems) [inline src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline]: I still disagree with extraneous newlines, IMO there's no loss of ability to navigate the code.

File: src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline [hunk 5]
```diff
@@ -1,871 +1,966 @@
...
 
+    Logging loggingService = createService();
```
Comment (saicheems) [inline src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline]: #556

File: src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline (hunk 2 above)
Comment (saicheems) [inline src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline]: #556

File: src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline (hunk 1 above)
Comment (saicheems) [inline src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline]: #556

File: src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline [hunk 6]
```diff
@@ -1,871 +1,966 @@
...
-      requestBody.setPageToken(response.getNextPageToken());
     } while (response.getNextPageToken() != null);
   }
+
+  public Logging createService() throws IOException, GeneralSecurityException {
+    HttpTransport httpTransport = GoogleNetHttpTransport.newTrustedTransport();
+    JsonFactory jsonFactory = JacksonFactory.getDefaultInstance();
+
+    GoogleCredential credential = GoogleCredential.getApplicationDefault();
+
```
Comment (saicheems) [inline src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline]: #556

File: src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline [hunk 7]
```diff
@@ -1,871 +1,966 @@
...
+      credential =
+          credential.createScoped(Arrays.asList("https://www.googleapis.com/auth/cloud-platform"));
+    }
+
+    return new Logging.Builder(httpTransport, jsonFactory, credential)
+        .setApplicationName("Google Cloud Platform Sample")
+        .build();
+  }
+
+  public static void main(String[] args) throws IOException, GeneralSecurityException {
```
Comment (saicheems) [inline src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline]: #556

File: src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline [hunk 8]
```diff
@@ -1,871 +1,966 @@
...
+    }
 
-    Logging.Entries.Write request = loggingService.entries().write(requestBody);
-    WriteLogEntriesResponse response = request.execute();
+    return new Logging.Builder(httpTransport, jsonFactory, credential)
+        .setApplicationName("Google Cloud Platform Sample")
+        .build();
+  }
 
-    // TODO: Change code below to process the `response` object:
```
Comment (saicheems) [inline src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline]: #556

File: src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline [hunk 9]
```diff
@@ -1,871 +1,966 @@
...
-    // Use Application Default Credentials for authentication when running locally.
-    // For more information, see:
-    // https://developers.google.com/identity/protocols/application-default-credentials
-    GoogleCredential credential = GoogleCredential.getApplicationDefault();
-    if (credential.createScopedRequired()) {
-      credential = credential.createScoped(Collections.singletonList("https://www.googleapis.com/auth/cloud-platform"));
-    }
+  public void run() throws IOException, GeneralSecurityException {
+    // Required. The resource name of the log to delete. Example: `"projects/my-project/logs/syslog"`.
+    String logName = "projects/{MY-PROJECT}/logs/{MY-LOG}"; // TODO: Update placeholder values.
```
Comment (saicheems) [inline src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline]: #556

File: src/test/java/com/google/api/codegen/testdata/discoveries/java/java_translate.v2.json.baseline [hunk 10]
```diff
@@ -10,53 +10,54 @@
...
-      .setApplicationName("Google Cloud Platform Sample")
-      .build();
-
-
-    // TODO: Change placeholder below to desired parameter value for the `list` method:
-
+  public void run() throws IOException, GeneralSecurityException {
     // The text to detect
-    List<String> q = new ArrayList<String>();
+    java.util.List<String> q = new ArrayList<>(); // TODO: Update placeholder value.
```
Comment (saicheems) [inline src/test/java/com/google/api/codegen/testdata/discoveries/java/java_translate.v2.json.baseline]: It was a trade-off, admittedly, I was aware of this possibility case and I do still prefer it
The obvious fix for this is to fully qualify the request type and not the list type, but I've changed it back anyway

#556

File: src/test/java/com/google/api/codegen/testdata/discoveries/java/java_translate.v2.json.baseline [hunk 11]
```diff
@@ -10,53 +10,54 @@
...
   }
 
-  public static Credential getAuth() {
-    // TODO: Change placeholder below to get authentication credentials.
+  public Translate createService() throws IOException, GeneralSecurityException {
+    HttpTransport httpTransport = GoogleNetHttpTransport.newTrustedTransport();
+    JsonFactory jsonFactory = JacksonFactory.getDefaultInstance();
+
+    // TODO: Change placeholder below to generate authentication credentials.
     // See: https://foo.com/bar
```
Comment (saicheems) [inline src/test/java/com/google/api/codegen/testdata/discoveries/java/java_translate.v2.json.baseline]: #556

**Comment by saicheems** [comment]
> sample.snip:8: Looks like this switch could be an if.
It could be, but the switch is cleaner IMO.

> sample.snip:48: Looks like example values for special patterns are missing.
I assume you're referring to values for `zone`... it's tricky, I'm not actually sure how to do it in MVVM...
Right now we're hardcoding the names of the APIs where there is a `zone` field to insert extra documentation. I have ideas around overrides, but the way this whole thing is designed to handle _special_ fields like that one is to use overrides to specify them.

Theoretically, that could mean a `sample()` field in `FieldInfo`, but those would have to be manually specified (or tool-assisted), and I assume that there will be negative feedback for that approach.

TLDR; I don't know how to do it nicely without manually specifying it
(If I had infinite power, I would change the API to include the sample in the docstring)

**Comment by tcoffee-google** [comment]
> > sample.snip:48: Looks like example values for special patterns are missing.
> They're there - `defaultValue` is populated with path template patterns if they're available.

I think we're talking about different things --- I'm talking about the additional comments currently implemented by `getDefaultString` in DiscoveryContext.java for date-times and the like.

File: src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline (hunk 4 above)
Comment (tcoffee-google) [inline src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline]: OK.

**Comment by tcoffee-google** [comment]
With exception of preceding comment, I've made all follow-ups on #556.

File: src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline (hunk 7 above)
Comment (vchudnov-g) [inline src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline]: I agree with @tcoffee-google. Thanks for addressing this in #556.

**Comment by saicheems** [comment]
OK, everything was addressed in #556 - rebased and ready for review.

File: src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline [hunk 12]
```diff
@@ -23,36 +22,34 @@ import com.google.api.client.json.jackson2.JacksonFactory;
...
-    // https://developers.google.com/identity/protocols/application-default-credentials
-    GoogleCredential credential = GoogleCredential.getApplicationDefault();
-    if (credential.createScopedRequired()) {
-      credential = credential.createScoped(Collections.singletonList("https://www.googleapis.com/auth/cloud-platform"));
-    }
+  public static void main(String args[]) throws IOException, GeneralSecurityException {
+    // Required. The resource name of the log to delete.  Example:
+    // `"projects/my-project/logs/syslog"`.
+    String logName = "";  // TODO: Update placeholder value.
+    // ex: "billingAccounts/my-billingAccount/logs/my-log"
```
Comment (tcoffee-google) [inline src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline]: This example doesn't seem to match the doc (?).

File: src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline [hunk 13]
```diff
@@ -320,36 +257,34 @@ import com.google.api.client.json.jackson2.JacksonFactory;
...
-    // https://developers.google.com/identity/protocols/application-default-credentials
-    GoogleCredential credential = GoogleCredential.getApplicationDefault();
-    if (credential.createScopedRequired()) {
-      credential = credential.createScoped(Collections.singletonList("https://www.googleapis.com/auth/cloud-platform"));
-    }
+  public static void main(String args[]) throws IOException, GeneralSecurityException {
+    // Required. The resource name of the log to delete.  Example:
+    // `"projects/my-project/logs/syslog"`.
+    String logName = "";  // TODO: Update placeholder value.
+    // ex: "organizations/my-organization/logs/my-log"
```
Comment (tcoffee-google) [inline src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline]: Another mismatch I noticed.

File: src/test/java/com/google/api/codegen/testdata/discoveries/java/java_translate.v2.json.baseline [hunk 14]
```diff
@@ -16,37 +16,27 @@ import com.google.api.client.http.HttpTransport;
...
     DetectionsListResponse response = request.execute();
+
+    // TODO: Change code below to process the `response` object:
     System.out.println(response);
   }
 
-  public Translate createService() throws IOException, GeneralSecurityException {
+  public static Translate createTranslateService() throws IOException, GeneralSecurityException {
     HttpTransport httpTransport = GoogleNetHttpTransport.newTrustedTransport();
     JsonFactory jsonFactory = JacksonFactory.getDefaultInstance();
```
Comment (tcoffee-google) [inline src/test/java/com/google/api/codegen/testdata/discoveries/java/java_translate.v2.json.baseline]: Are we going to follow formatting elsewhere and put "See:" on the previous line, below?

Comment (saicheems) [inline src/test/java/com/google/api/codegen/testdata/discoveries/java/java_translate.v2.json.baseline]: This comment seems to have disappeared in relation to the baseline... so I assume you're talking about the same issue that #570 is resolving, so yes.

File: src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline (hunk 12 above)
Comment (tcoffee-google) [inline src/test/java/com/google/api/codegen/testdata/discoveries/java/java_logging.v2beta1.json.baseline]: OK, I suspected that might be the case. File a bug against the API doc?

**Comment by saicheems** [comment]
Rebased and re-ready

**Comment by vchudnov-g** [comment]
If the compilecheck passes with this new flow, then LGTM--but please wait for @tcoffee-google's LGTM as well.

**Comment by tcoffee-google** [comment]
Sorry, one more nit: now that the "Assign values to desired fields of `requestBody`" comment appears above the `requestBody` variable declaration, let's end this comment with a colon instead of a period.

Otherwise LGTM.
